File: stack_text_editor.py
import os

class DynamicStackBasedTextEditor:
    def __init__(self):
        """
        Initializes the text editor with an empty string and two stacks:
        - undo_stack: to store actions for undo
        - redo_stack: to store actions for redo
        """
        self.text = ""
        self.undo_stack = []  
        self.redo_stack = []  
        self.current_file = None  

    def create_file(self, file_path):
        """
        Creates a new file if it doesn't exist and sets it as the current file.
        
        Parameters:
            file_path (str): Path to the file to be created.
        """
        if os.path.exists(file_path):
            print(f"File '{file_path}' already exists.")
        else:
            with open(file_path, 'w') as file:
                pass  
            self.current_file = file_path
            self.text = ""
            self.undo_stack.clear()
            self.redo_stack.clear()
            print(f"New file '{file_path}' created and loaded.")

    def load_file(self, file_path):
        if os.path.exists(file_path):
            with open(file_path, 'r') as file:
                self.text = file.read()
            self.current_file = file_path
            print(f"Loaded text from '{file_path}':\n{self.text}")
        else:
            print(f"File '{file_path}' not found.")

    def save_file(self, file_path=None):
        if file_path:
            self.current_file = file_path
        
        if self.current_file:
            with open(self.current_file, 'w') as file:
                file.write(self.text)
            print(f"Text saved to '{self.current_file}'")
        else:
            print("No file specified. Use 'save <filename>' to specify a file.")

    def insert(self, new_text):
        if self.text:
            new_text = " " + new_text

        self.undo_stack.append(("insert", new_text))
        self.redo_stack.clear()
        self.text += new_text
        print(f"Inserted: '{new_text}' -> Current Text: '{self.text}'")

    def delete_text(self, text_to_delete):
        if text_to_delete in self.text:
            occurrences = [pos for pos in range(len(self.text)) if self.text.startswith(text_to_delete, pos)]
            self.undo_stack.append(("delete_text", text_to_delete, occurrences))
            self.redo_stack.clear()
            
            self.text = self.text.replace(text_to_delete, "")
            print(f"Deleted all instances of: '{text_to_delete}' -> Current Text: '{self.text}'")
        else:
            print(f"Text '{text_to_delete}' not found in current text.")

    def undo(self):
        if not self.undo_stack:
            print("Nothing to undo!")
            return

        action = self.undo_stack.pop()
        
        if action[0] == "insert":
            content = action[1]
            self.text = self.text[:-len(content)]
            self.redo_stack.append(("insert", content))
            print(f"Undo Insert -> Current Text: '{self.text}'")
            
        elif action[0] == "delete_text":
            text_to_insert, positions = action[1], action[2]
            for pos in reversed(positions):
                self.text = self.text[:pos] + text_to_insert + self.text[pos:]
            self.redo_stack.append(("delete_text", text_to_insert, positions))
            print(f"Undo Delete Text -> Current Text: '{self.text}'")

    def redo(self):
        if not self.redo_stack:
            print("Nothing to redo!")
            return

        action = self.redo_stack.pop()
        
        if action[0] == "insert":
            content = action[1]
            self.text += content
            self.undo_stack.append(("insert", content))
            print(f"Redo Insert -> Current Text: '{self.text}'")
            
        elif action[0] == "delete_text":
            text_to_delete, positions = action[1], action[2]
            self.text = self.text.replace(text_to_delete, "")
            self.undo_stack.append(("delete_text", text_to_delete, positions))
            print(f"Redo Delete Text -> Current Text: '{self.text}'")

    def display_text(self):
        print(f"Current Text: '{self.text}'")


def main():
    editor = DynamicStackBasedTextEditor()
    
    print("Welcome to the Stack-Based Text Editor.")
    print("Commands: create <file>, load <file>, save <file>, insert <text>, delete_text <text>, undo, redo, display, quit")

    while True:
        command = input("Enter command: ").strip().split(" ", 1)
        action = command[0].lower()

        if action == "create" and len(command) > 1:
            file_path = command[1]
            editor.create_file(file_path)

        elif action == "load" and len(command) > 1:
            file_path = command[1]
            editor.load_file(file_path)

        elif action == "save" and len(command) > 1:
            file_path = command[1]
            editor.save_file(file_path)
        
        elif action == "save" and len(command) == 1:
            editor.save_file()

        elif action == "insert" and len(command) > 1:
            editor.insert(command[1])

        elif action == "delete_text" and len(command) > 1:
            editor.delete_text(command[1])

        elif action == "undo":
            editor.undo()

        elif action == "redo":
            editor.redo()

        elif action == "display":
            editor.display_text()

        elif action == "quit":
            break

        else:
            print("Invalid command. Try again.")


if __name__ == "__main__":
    main()

File: stack_text_editor.cpp
#include <iostream>
#include <fstream>
#include <stack>
#include <string>
#include <vector>
#include <algorithm>
#include <filesystem>

class DynamicStackBasedTextEditor
{
private:
    std::string text;
    std::stack<std::pair<std::string, std::string>> undoStack;
    std::stack<std::pair<std::string, std::string>> redoStack;
    std::string currentFile;

    std::vector<size_t> findOccurrences(const std::string &target)
    {
        std::vector<size_t> positions;
        size_t pos = text.find(target);
        while (pos != std::string::npos)
        {
            positions.push_back(pos);
            pos = text.find(target, pos + 1);
        }
        return positions;
    }

public:
    DynamicStackBasedTextEditor() : text(""), currentFile("") {}

    void createFile(const std::string &filePath)
    {
        if (std::filesystem::exists(filePath))
        {
            std::cout << "File '" << filePath << "' already exists.\n";
        }
        else
        {
            std::ofstream file(filePath);
            file.close();
            currentFile = filePath;
            text.clear();
            while (!undoStack.empty())
                undoStack.pop();
            while (!redoStack.empty())
                redoStack.pop();
            std::cout << "New file '" << filePath << "' created and loaded.\n";
        }
    }

    void loadFile(const std::string &filePath)
    {
        if (std::filesystem::exists(filePath))
        {
            std::ifstream file(filePath);
            text.assign((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
            currentFile = filePath;
            std::cout << "Loaded text from '" << filePath << "':\n"
                      << text << "\n";
        }
        else
        {
            std::cout << "File '" << filePath << "' not found.\n";
        }
    }

    void saveFile(const std::string &filePath = "")
    {
        if (!filePath.empty())
        {
            currentFile = filePath;
        }

        if (!currentFile.empty())
        {
            std::ofstream file(currentFile);
            file << text;
            file.close();
            std::cout << "Text saved to '" << currentFile << "'\n";
        }
        else
        {
            std::cout << "No file specified. Use 'save <filename>' to specify a file.\n";
        }
    }

    void insert(const std::string &newText)
    {
        std::string insertText = (text.empty() ? "" : " ") + newText;
        undoStack.push({"insert", insertText});
        while (!redoStack.empty())
            redoStack.pop();
        text += insertText;
        std::cout << "Inserted: '" << insertText << "' -> Current Text: '" << text << "'\n";
    }

    void deleteText(const std::string &textToDelete)
    {
        auto positions = findOccurrences(textToDelete);
        if (!positions.empty())
        {
            undoStack.push({"delete", textToDelete});
            while (!redoStack.empty())
                redoStack.pop();
            text.erase(std::remove(text.begin(), text.end(), textToDelete), text.end());
            std::cout << "Deleted all instances of: '" << textToDelete << "' -> Current Text: '" << text << "'\n";
        }
        else
        {
            std::cout << "Text '" << textToDelete << "' not found in current text.\n";
        }
    }

    void undo()
    {
        if (undoStack.empty())
        {
            std::cout << "Nothing to undo!\n";
            return;
        }

        auto action = undoStack.top();
        undoStack.pop();

        if (action.first == "insert")
        {
            text.erase(text.size() - action.second.size());
            redoStack.push(action);
            std::cout << "Undo Insert -> Current Text: '" << text << "'\n";
        }
        else if (action.first == "delete")
        {
            text += action.second;
            redoStack.push(action);
            std::cout << "Undo Delete -> Current Text: '" << text << "'\n";
        }
    }

    void redo()
    {
        if (redoStack.empty())
        {
            std::cout << "Nothing to redo!\n";
            return;
        }

        auto action = redoStack.top();
        redoStack.pop();

        if (action.first == "insert")
        {
            text += action.second;
            undoStack.push(action);
            std::cout << "Redo Insert -> Current Text: '" << text << "'\n";
        }
        else if (action.first == "delete")
        {
            text.erase(std::remove(text.begin(), text.end(), action.second), text.end());
            undoStack.push(action);
            std::cout << "Redo Delete -> Current Text: '" << text << "'\n";
        }
    }

    void displayText()
    {
        std::cout << "Current Text: '" << text << "'\n";
    }
};

int main()
{
    DynamicStackBasedTextEditor editor;
    std::cout << "Welcome to the Stack-Based Text Editor.\n";
    std::cout << "Commands: create <file>, load <file>, save <file>, insert <text>, delete <text>, undo, redo, display, quit\n";

    std::string command;
    while (true)
    {
        std::cout << "Enter command: ";
        std::getline(std::cin, command);

        if (command.rfind("create", 0) == 0)
        {
            editor.createFile(command.substr(7));
        }
        else if (command.rfind("load", 0) == 0)
        {
            editor.loadFile(command.substr(5));
        }
        else if (command.rfind("save", 0) == 0)
        {
            editor.saveFile(command.substr(5));
        }
        else if (command.rfind("insert", 0) == 0)
        {
            editor.insert(command.substr(7));
        }
        else if (command.rfind("delete", 0) == 0)
        {
            editor.deleteText(command.substr(7));
        }
        else if (command == "undo")
        {
            editor.undo();
        }
        else if (command == "redo")
        {
            editor.redo();
        }
        else if (command == "display")
        {
            editor.displayText();
        }
        else if (command == "quit")
        {
            break;
        }
        else
        {
            std::cout << "Invalid command. Try again.\n";
        }
    }

    return 0;
}
